src/quadtree.cpp

```diff
@@ -0,0 +1,199 @@
+#include "quadtree.h"
+
+namespace {
+std::unique_ptr<QuadTree> nodes[4] = {};
+
+uint32_t create_index(uint16_t x, uint16_t y) { return ((x & 0x8000) >> 15) | ((y & 0x8000) >> 14); }
+
+Leaf* find_leaf_in_root(uint16_t x, uint16_t y)
+{
+	auto index = create_index(x, y);
+	if (auto next_node = nodes[index].get()) {
+		return find_leaf(x, y, next_node);
+	}
+	return nullptr;
+}
+
+Leaf* find_leaf(uint16_t x, uint16_t y, QuadTree* node)
+{
+	if (node->is_leaf()) {
+		return static_cast<Leaf*>(node);
+	}
+
+	auto index = create_index(x, y);
+	if (auto next_node = node->nodes[index].get()) {
+		return find_leaf(x << 1, y << 1, next_node);
+	}
+	return nullptr;
+}
+
+void create_leaf_in_root(uint16_t x, uint16_t y, uint8_t z)
+{
+	auto index = create_index(x, y);
+	if (!nodes[index]) {
+		nodes[index] = std::make_unique<Node>();
+	}
+
+	create_leaf(x, y, (MAP_MAX_LAYERS - 1), nodes[index].get());
+}
+
+void create_leaf(uint16_t x, uint16_t y, uint8_t z, QuadTree* node)
+{
+	if (node->is_leaf()) {
+		return;
+	}
+
+	auto index = create_index(x, y);
+	if (!node->nodes[index]) {
+		if (z == FLOOR_BITS) {
+			node->nodes[index] = std::make_unique<Leaf>(x, y);
+		} else {
+			node->nodes[index] = std::make_unique<Node>();
+		}
+	}
+
+	create_leaf(x * 2, y * 2, z - 1, node->nodes[index].get());
+}
+
+} // namespace
+
+void tfs::quadtree::find(uint16_t start_x, uint16_t start_y, uint16_t end_x, uint16_t end_y, SpectatorVec& spectators,
+                         std::function<bool(Creature*S)> comparasion)
```

Comment:
~~Comparación~~ `comparison`
```suggestion
                         std::function<bool(Creature*)> comparison)
```

Or just `cmp`

Reply:
I was thinking of something like `std::function<bool(T)>` and `quadtree::find<T*>(...)`. I did some tests, but I will leave it for after we finish base quadtree.